### PR TITLE
MNT: overhaul stale handling

### LINKED ISSUE #4732: Interactivity is fragile
With the current master, I am finding that interactivity can break in case of errors.
This is one example:
```python
>>> import matplotlib
>>> matplotlib.use("tkagg")
>>> from pylab import *
>>> figure()
<matplotlib.figure.Figure object at 0x103b03150>
>>> line, = plot([1,2,3])
>>> line
<matplotlib.lines.Line2D object at 0x108f81350>
>>> line.set_dashes
<bound method Line2D.set_dashes of <matplotlib.lines.Line2D object at 0x108f81350>>
>>> line.set_dashes('--') # incorrect usage of set_dashes
>>> Exception in Tkinter callback
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk/Tkinter.py", line 1536, in __call__
    return self.func(*args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk/Tkinter.py", line 587, in callit
    func(*args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/backends/backend_tkagg.py", line 370, in idle_draw
    self.draw()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/backends/backend_tkagg.py", line 354, in draw
    FigureCanvasAgg.draw(self)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/backends/backend_agg.py", line 471, in draw
    self.figure.draw(self.renderer)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/figure.py", line 1119, in draw
    func(*args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/axes/_base.py", line 2169, in draw
    a.draw(renderer)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/artist.py", line 60, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/lines.py", line 741, in draw
    drawFunc(renderer, gc, tpath, affine.frozen())
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/lines.py", line 1154, in _draw_lines
    self._lineFunc(renderer, gc, path, trans)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/lines.py", line 1201, in _draw_dashed
    renderer.draw_path(gc, path, trans)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-macosx-10.9-x86_64.egg/matplotlib/backends/backend_agg.py", line 166, in draw_path
    self._renderer.draw_path(gc, path, transform, rgbFace)
TypeError: a float is required

>>> plot([1,2,4])
# nothing happens

>>> cla()
# nothing happens
```
Resizing the figure restores interactivity.

### PR DESCRIPTION
**Pull request by tacaswell**
Closes #4732

### CODE REVIEW

**Comment by WeatherGod** [comment]
This PR actually only has a single failure (the same in all python versions):
```
======================================================================
FAIL: matplotlib.tests.test_offsetbox.test_offsetbox_clip_children
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.6/site-packages/matplotlib-1.5.dev1-py2.6-linux-x86_64.egg/matplotlib/testing/decorators.py", line 110, in wrapped_function
    func(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.6/site-packages/matplotlib-1.5.dev1-py2.6-linux-x86_64.egg/matplotlib/tests/test_offsetbox.py", line 83, in test_offsetbox_clip_children
    assert_true(fig.stale)
AssertionError
```

**Comment by tacaswell** [comment]
Turns out the issue was that funny things were happening with the `offsetbox` subclasses which was getting masked by the fact that before artists were pinging both the axes and the figure, now they just hit the figure.

There are probably a bunch more issues like this that we will slowly find...

**Comment by tacaswell** [comment]
@rmorshea  the stale callbacks have been simplified to not use `pchanged` anymore. I think this will make replicating the current proposed behavior easier.

**Comment by mdboom** [comment]
I definitely see this as an improvement, modulo investigating the Travis failure.  Possibly more to do over time, but I don't see any reason not to do this much now.

**Comment by tacaswell** [comment]
The travis issue is triggering `draw`s in a `draw`.  This is caused by:

 - not checking if anything actually changed, just setting stale on every setter so stale gets triggered during draws
 - the stale state now propagating past existing `stale = True` states
 - `Agg` not having an asynchronous draw method and `draw_idle` just passing through to draw.

@mdehoon makes the very good point that due to our late validation of values if you set an invalid value you don't find out until draw time.  If the stale state does not propagate past the existing stale artists and you are using plain python then a re-draw won't be triggered the next time the value is changed (as implemented with IPython it will work fine as-is). 

The solutions to this that I see are:

 - put back in the short circuit logic back in and tell people not to use plain python (I am +0 on this, and @mdehoon is -1)
 - do proper validation and change determination in ever setter (I am -1 on this as traitlets will take care of this)
 - make Agg's draw_idle re-entrant (I am +0 and am guessing @mdehoon is +1)

I am working on adding a commit to make the base `draw_idle` re-entrant.

**Comment by tacaswell** [comment]
I think this is ready to go again.

File: examples/pylab_examples/system_monitor.py [hunk 1]
```diff
@@ -44,10 +44,21 @@ def get_stats():
...
     pn.set_height(n)
+
+    # ask the canvas to re-draw itself the next time it
+    # has a chance.
+    # For most of the GUI backends this adds an event to the queue
+    # of the GUI frame works event loop.
+    fig.canvas.draw_idle()
     try:
+        # make sure that the GUI framework has a chance to run it's event loop
+        # and clear and GUI events.  This needs to be in a try/except block
```
Comment (WeatherGod) [inline examples/pylab_examples/system_monitor.py]: "frame works" --> "framework's"
"it's" --> "its"
"and" --> "any"

Comment (tacaswell) [inline examples/pylab_examples/system_monitor.py]: Yikes.

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -2059,7 +2059,8 @@ def draw(self, renderer=None, inframe=False):
         if not self.get_visible():
             return
         renderer.open_group('axes')
-
+        # prevent triggering call backs during the draw process
+        self._stale = True
```
Comment (WeatherGod) [inline lib/matplotlib/axes/_base.py]: not sure I understand why drawing would need to set stale flag to True.

Comment (tacaswell) [inline lib/matplotlib/axes/_base.py]: This can probably be removed now that stale state _always_ propagates (which I am not 100% happy with) and with the context manager in `draw_idle`.  As I babbled in the comments, the stale code is very aggressive about setting artists to be stale (even if it hasn't changed).  In the case of `Agg` this was causing redraws to be trigged from inside of a draw.

File: lib/matplotlib/offsetbox.py [hunk 3]
```diff
@@ -187,6 +187,14 @@ def set_figure(self, fig):
         for c in self.get_children():
             c.set_figure(fig)
 
+    @martist.Artist.axes.setter
+    def axes(self, ax):
+        # TODO deal with this better
+        martist.Artist.axes.fset(self, ax)
```
Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: uhm, never seen this before, and the documentation is nil. What does this do?

Comment (tacaswell) [inline lib/matplotlib/offsetbox.py]: This is how you call the base-class property setter function.

https://docs.python.org/2/howto/descriptor.html#properties

**Comment by tacaswell** [comment]
The switch away from `None` was @mdboom 's idea in a comment on #4849

File: lib/matplotlib/offsetbox.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: heh, so that's how you do that... good to know!

**Comment by tacaswell** [comment]
The handling of the stale call back could also be done by using a fixed function and not stripping it out in `__getstate__`.  This is simpler on ourside, but more limiting if users want to do pickling + crazy custom stale callbacks.  Not sure if we care about that.

Traitlets are going to overhaul all of this anyway so it is probably not worth too much effort.

**Comment by tacaswell** [comment]
and I think this one is ready to go

**Comment by tacaswell** [comment]
@jrevans The top commits on this fix animation + blitting 

attn @dopplershift Can you make sure what I did make sense?

**Comment by dopplershift** [comment]
:+1: to the changes for animation. Those all make sense--makes me wonder why I didn't do that in the first place. :)

File: lib/matplotlib/animation.py [hunk 4]
```diff
@@ -971,17 +971,18 @@ def __init__(self, fig, artists, *args, **kwargs):
 
     def _init_draw(self):
         # Make all the artists involved in *any* frame invisible
-        axes = []
+        figs = set()
```
Comment (WeatherGod) [inline lib/matplotlib/animation.py]: wait a minute... figures are mutable. How is it that we can have a set of them?

Comment (tacaswell) [inline lib/matplotlib/animation.py]: Good catch on that and because i bet it is defaulting to using the object id

**Comment by dopplershift** [comment]
Dunno, this just worked for me:
```

In [1]: class A(object):
   ...:     pass
   ...:

In [2]: a = A()

In [3]: set([a])
Out[3]: {<__main__.A at 0x103841208>}

In [4]: a
Out[4]: <__main__.A at 0x103841208>

In [5]: s = set([a])

In [6]: a
Out[6]: <__main__.A at 0x103841208>

In [7]: a.foo = 1

In [8]: a.bar = 2

In [9]: a in s
Out[9]: True
```

File: lib/matplotlib/animation.py [hunk 5]
```diff
@@ -1105,3 +1108,5 @@ def _draw_frame(self, framedata):
         # Call the func with framedata and args. If blitting is desired,
         # func needs to return a sequence of any artists that were modified.
         self._drawn_artists = self._func(framedata, *self._args)
+        for a in self._drawn_artists:
+            a.set_animated(self._blit)
```
Comment (WeatherGod) [inline lib/matplotlib/animation.py]: Isn't this going to wreak havoc on the stale-handling code? Don't we want to avoid setting artists as stale within drawing code paths?

**Comment by dopplershift** [comment]
Ah, see: https://docs.python.org/2/glossary.html#term-hashable (Python 2) and https://docs.python.org/3/glossary.html#term-hashable (Python 3).

TLDR; basically using `id()`.

File: lib/matplotlib/artist.py [hunk 6]
```diff
@@ -236,15 +235,16 @@ def stale(self):
...
-            # only trigger propagation if marking as stale
-            if self._stale:
-                self.pchanged()
+        self._stale = val
+
+        # if the artist is animated it does not take normal part in the
+        # draw stack and is not expected to be drawn as part of the normal
+        # draw loop (when not saving) so do not propagate this change
+        if self.get_animated():
+            return
```
Comment (WeatherGod) [inline lib/matplotlib/artist.py]: Is it still going to be possible to change the properties of an animated object? I typically do things like changing the alpha or color as part of an animation.

Comment (dopplershift) [inline lib/matplotlib/artist.py]: IIRC, the animated bit on an `Artist` essentially means "I need to manage how this is redrawn, so don't do it for me." So everything should just work as normal, this just makes blitting work a bit better.

**Comment by WeatherGod** [comment]
Still, I don't see the value in using sets here, and I don't like that it works now on a technicality. A list would work just fine. It isn't like we are expecting to have many figures to check against. At the extreme end, i would imagine someone doing something fancy with 10 figures, but typically, only 1 or 3 figures.

File: lib/matplotlib/artist.py (hunk 6 above)
Comment (tacaswell) [inline lib/matplotlib/artist.py]: The stale handler does not propagate on animated artists, but you can still
set all of their properties.  If you are using blitting you are still
responsible for doing the redraw. If you are not blitting then animated
needs to be false so they get drawn by the normal draw method which means
stale state does propagate, but draw_idle takes care of collecting those.

I am a bit worried about saving, i suspect that should also be checked in
the figure call back installed by pyplot.

Comment (tacaswell) [inline lib/matplotlib/artist.py]: I will revert the set change when i get back to my normal computer.

**Comment by dopplershift** [comment]
I disagree about it working on a "technicality"--this is documented python behavior, essentially making the `in` check work using `is`. In fact, using `in` (which uses `__eq__`) ends up on the same code path as `hash`:
```
User-defined classes have __eq__() and __hash__() methods by default; with them, all objects compare unequal (except with themselves) and x.__hash__() returns an appropriate value such that x == y implies both that x is y and hash(x) == hash(y).
```
(https://docs.python.org/3/reference/datamodel.html#object.__eq__)

However, I do agree that there is essentially no performance difference in our real-world scenarios here.

**Comment by WeatherGod** [comment]
Ok, just tried out this branch. I was doing an imshow of a very large array (3600x7000), and a basemap plot and using the TkAgg backend. Watching "top" in another terminal, I see the memory usage increasing every time my mouse passes over the figure window. It never goes back down. The amount of memory increases by hundreds of MB for a quick pass over the window.

I don't know yet if it is because of this PR, or if it is due to some unfixed issue with stale flags, or something else entirely. Just noting it here for the moment.

**Comment by WeatherGod** [comment]
Joys... this is on master and this branch. Furthermore, once you pass over the figure window, the memory usage keeps creeping up, even though the mouse is no longer over the window.

**Comment by tacaswell** [comment]
Try commenting out the mouse hover logic in backed base.

**Comment by WeatherGod** [comment]
Yup, looks like commenting out the contents of `NavigationToolbar2.mouse_move` "fixed" the problem. Do you think it is a bug to be dealt with here or shall I file a new issue?

**Comment by tacaswell** [comment]
Tack it on to the pr where i am working on removing hit list

**Comment by jenshnielsen** [comment]
I can still reproduce the original issue in #4732 when interactivity is on i.e. `plt.ion()` 
However I can also reproduce the issue on 1.4.3 so I not really cure what to make of this

**Comment by jenshnielsen** [comment]
I.e if I call `set_dashes('--')` it  triggers that exception on every draw until call `set_dashes((1,1))` or whatever but surely that is just missing input validation?

**Comment by tacaswell** [comment]
That is the expected behavior, the problem was that previously correcting
the value would not trigger a redraw at all and the was no obvious way to
do so.

**Comment by jenshnielsen** [comment]
@tacaswell That makes sense. However, I can't reproduce that on current master. 

`set_dashes((3,'3))` after `set_dashes(('--'))` correctly triggers the redraw so it seems that either this has already been fixed by something else or I am doing something stupid. In any case this is an improvement to the stale login so I will go ahead and merge it

**Comment by tacaswell** [comment]
Are you using ipython or plain python? The redraw logic is different in the
two cases.

**Comment by jenshnielsen** [comment]
plain python
